### PR TITLE
[Ready] Port moveit_fake_controller_manager to ROS2

### PR DESCRIPTION
**Pull request by ddengster**
### Description
Port moveit_fake_controller_manager to ROS2

The remaining warning is from srdfdom, wont fix

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)

### CODE REVIEW

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp [hunk 1]
```diff
@@ -74,19 +74,19 @@ LastPointController::~LastPointController() = default;
...
   if (t.joint_trajectory.points.empty())
     return true;
 
-  sensor_msgs::JointState js;
-  const trajectory_msgs::JointTrajectoryPoint& last = t.joint_trajectory.points.back();
+  sensor_msgs::msg::JointState js;
+  const trajectory_msgs::msg::JointTrajectoryPoint& last = t.joint_trajectory.points.back();
   js.header = t.joint_trajectory.header;
-  js.header.stamp = ros::Time::now();
+  js.header.stamp = rclcpp::Clock().now();
```
Comment (henningkayser) [inline moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp:84]: Please use `RCL_ROS_TIME` for the clock

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp [hunk 2]
```diff
@@ -158,54 +159,56 @@ ViaPointController::~ViaPointController() = default;
...
+  sensor_msgs::msg::JointState js;
   js.header = t.joint_trajectory.header;
   js.name = t.joint_trajectory.joint_names;
 
   // publish joint states for all intermediate via points of the trajectory
   // no further interpolation
-  ros::Time start_time = ros::Time::now();
-  for (std::vector<trajectory_msgs::JointTrajectoryPoint>::const_iterator via = t.joint_trajectory.points.begin(),
-                                                                          end = t.joint_trajectory.points.end();
+  rclcpp::Time start_time = rclcpp::Clock().now();
```
Comment (henningkayser) [inline moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp:169]: same

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.h [hunk 3]
```diff
@@ -34,16 +34,15 @@
  *********************************************************************/
 
 /* Author: Ioan Sucan, Robert Haschke */
+#ifndef MOVEIT_FAKE_CONTROLLERS
```
Comment (henningkayser) [inline moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.h:37]: Better use `#pragma once`

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp [hunk 4]
```diff
@@ -35,167 +35,173 @@
...
 public:
-  MoveItFakeControllerManager() : node_handle_("~")
+  MoveItFakeControllerManager() : MoveItControllerManager()
   {
-    if (!node_handle_.hasParam("controller_list"))
+  }
+
+  void initialize(const rclcpp::Node::SharedPtr& node)
+  {
+    // TODO(henningkayser): use flexible base
```
Comment (henningkayser) [inline moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp:60]: Do we need to address this TODO here? I would imagine that a variable namespace is configured using the node interfaces.

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp [hunk 5]
```diff
@@ -35,167 +35,173 @@
...
 
     /* publish initial pose */
-    XmlRpc::XmlRpcValue initial;
-    if (node_handle_.getParam("initial", initial))
-    {
-      sensor_msgs::JointState js = loadInitialJointValues(initial);
-      js.header.stamp = ros::Time::now();
-      pub_.publish(js);
-    }
+    // TODO: codebase wide refactoring for XmlRpc
```
Comment (henningkayser) [inline moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp:81]: I think we should add a basic implementation for this. Loading this parameter without XmlRpc shouldn't be too much an issue, no?

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp [hunk 6]
```diff
@@ -35,167 +35,173 @@
...
-        ROS_ERROR_NAMED("MoveItFakeControllerManager", "Caught unknown exception while parsing controller information");
+        RCLCPP_ERROR(LOGGER, "Caught unknown exception while parsing controller information");
       }
     }
   }
 
-  sensor_msgs::JointState loadInitialJointValues(XmlRpc::XmlRpcValue& param) const
-  {
-    sensor_msgs::JointState js;
+  // TODO: codebase wide refactoring for XmlRpc
```
Comment (henningkayser) [inline moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp:131]: We could easily work around missing XmlRpc support by structuring the parameters like this:

```
initial:
  <group>: <pose>
```

Since we know the robot model, it would be straight forward to just look for parameters using the group names and then just use the existing checks if `pose` is a valid named target. We don't need an array of structs if we already know that group names are unique identifiers and can be used as parameter keys.

File: moveit_plugins/moveit_fake_controller_manager/CMakeLists.txt [hunk 7]
```diff
@@ -9,34 +9,27 @@ if(NOT CMAKE_CONFIGURATION_TYPES AND NOT CMAKE_BUILD_TYPE)
...
+        LIBRARY DESTINATION lib
+        ARCHIVE DESTINATION lib
+        RUNTIME DESTINATION bin) 
+
+pluginlib_export_plugin_description_file(moveit_core moveit_fake_controller_manager_plugin_description.xml)
 
-install(FILES moveit_fake_controller_manager_plugin_description.xml
-        DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}
-       )
+ament_export_libraries(${PROJECT_NAME})
```
Comment (henningkayser) [inline moveit_plugins/moveit_fake_controller_manager/CMakeLists.txt:34]: Shouldn't we also add `ament_export_dependencies()`?

Comment (ddengster) [reply moveit_plugins/moveit_fake_controller_manager/CMakeLists.txt:34]: added f228b20

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.h (hunk 3 above)
Comment (ddengster) [reply moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.h]: added f228b20

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp (hunk 2 above)
Comment (ddengster) [reply moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp]: fixed f228b20

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp (hunk 1 above)
Comment (ddengster) [reply moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp]: fixed f228b20

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp (hunk 4 above)
Comment (ddengster) [reply moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp:60]: I'm not sure what you meant by flexible base, so I left it alone.

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp (hunk 6 above)
Comment (ddengster) [reply moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp:129]: ported e9b3c1d

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp (hunk 5 above)
Comment (ddengster) [reply moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp]: port e9b3c1d

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp [hunk 8]
```diff
@@ -158,54 +159,56 @@ ViaPointController::~ViaPointController() = default;
...
+                                                                               end = t.joint_trajectory.points.end();
        !cancelled() && via != end; ++via)
   {
     js.position = via->positions;
     js.velocity = via->velocities;
     js.effort = via->effort;
 
-    ros::Duration wait_time = via->time_from_start - (ros::Time::now() - start_time);
-    if (wait_time.toSec() > std::numeric_limits<float>::epsilon())
+    rclcpp::Duration wait_time = rclcpp::Duration(via->time_from_start) - (rclcpp::Clock().now() - start_time);
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp:178]:
```suggestion
    rclcpp::Duration wait_time = rclcpp::Duration(via->time_from_start) - (rclcpp::Clock(RCL_ROS_TIME).now() - start_time);
```

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp [hunk 9]
```diff
@@ -158,54 +159,56 @@ ViaPointController::~ViaPointController() = default;
...
     {
-      ROS_DEBUG("Fake execution: waiting %0.1fs for next via point, %ld remaining", wait_time.toSec(), end - via);
-      wait_time.sleep();
+      RCLCPP_DEBUG(LOGGER, "Fake execution: waiting %0.1fs for next via point, %ld remaining", wait_time.seconds(),
+                   end - via);
+      rclcpp::sleep_for(std::chrono::nanoseconds(wait_time.nanoseconds()));
     }
-    js.header.stamp = ros::Time::now();
-    pub_.publish(js);
+    js.header.stamp = rclcpp::Clock().now();
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp:185]:
```suggestion
    js.header.stamp = rclcpp::Clock(RCL_ROS_TIME).now();
```

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp [hunk 10]
```diff
@@ -217,23 +220,23 @@ void interpolate(sensor_msgs::JointState& js, const trajectory_msgs::JointTrajec
...
 
-  const std::vector<trajectory_msgs::JointTrajectoryPoint>& points = t.joint_trajectory.points;
-  std::vector<trajectory_msgs::JointTrajectoryPoint>::const_iterator prev = points.begin(),  // previous via point
+  const std::vector<trajectory_msgs::msg::JointTrajectoryPoint>& points = t.joint_trajectory.points;
+  std::vector<trajectory_msgs::msg::JointTrajectoryPoint>::const_iterator prev = points.begin(),  // previous via point
       next = points.begin() + 1,  // currently targetted via point
       end = points.end();
 
-  ros::Time start_time = ros::Time::now();
+  rclcpp::Time start_time = rclcpp::Clock().now();
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp:236]:
```suggestion
  rclcpp::Time start_time = rclcpp::Clock(RCL_ROS_TIME).now();
```

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp [hunk 11]
```diff
@@ -217,23 +220,23 @@ void interpolate(sensor_msgs::JointState& js, const trajectory_msgs::JointTrajec
...
   while (!cancelled())
   {
-    ros::Duration elapsed = ros::Time::now() - start_time;
+    rclcpp::Duration elapsed = rclcpp::Clock().now() - start_time;
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp:239]:
```suggestion
    rclcpp::Duration elapsed = rclcpp::Clock(RCL_ROS_TIME).now() - start_time;
```

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp [hunk 12]
```diff
@@ -243,29 +246,34 @@ void InterpolatingController::execTrajectory(const moveit_msgs::msg::RobotTrajec
...
+    double duration = (next_time - prev_time).seconds();
+    RCLCPP_DEBUG(LOGGER, "elapsed: %.3f via points %td,%td / %td  alpha: %.3f", elapsed.seconds(),
+                 prev - points.begin(), next - points.begin(), end - points.begin(),
+                 duration > std::numeric_limits<double>::epsilon() ?
+                     (elapsed - prev->time_from_start).seconds() / duration :
+                     1.0);
     interpolate(js, *prev, *next, elapsed);
-    js.header.stamp = ros::Time::now();
-    pub_.publish(js);
+    js.header.stamp = rclcpp::Clock().now();
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp:259]:
```suggestion
    js.header.stamp = rclcpp::Clock(RCL_ROS_TIME).now();
```

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp [hunk 13]
```diff
@@ -243,29 +246,34 @@ void InterpolatingController::execTrajectory(const moveit_msgs::msg::RobotTrajec
...
+
     rate_.sleep();
   }
   if (cancelled())
     return;
 
-  ros::Duration elapsed = ros::Time::now() - start_time;
-  ROS_DEBUG("elapsed: %.3f via points %td,%td / %td  alpha: 1.0", elapsed.toSec(), prev - points.begin(),
-            next - points.begin(), end - points.begin());
+  rclcpp::Duration elapsed = rclcpp::Clock().now() - start_time;
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp:267]:
```suggestion
  rclcpp::Duration elapsed = rclcpp::Clock(RCL_ROS_TIME).now() - start_time;
```

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp [hunk 14]
```diff
@@ -243,29 +246,34 @@ void InterpolatingController::execTrajectory(const moveit_msgs::msg::RobotTrajec
...
+  RCLCPP_DEBUG(LOGGER, "elapsed: %.3f via points %td,%td / %td  alpha: 1.0", elapsed.seconds(), prev - points.begin(),
+               next - points.begin(), end - points.begin());
 
   // publish last point
   interpolate(js, *prev, *prev, prev->time_from_start);
-  js.header.stamp = ros::Time::now();
-  pub_.publish(js);
+  js.header.stamp = rclcpp::Clock().now();
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.cpp:273]:
```suggestion
  js.header.stamp = rclcpp::Clock(RCL_ROS_TIME).now();
```

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.h [hunk 15]
```diff
@@ -113,15 +115,15 @@ class ViaPointController : public ThreadedController
...
   ~InterpolatingController() override;
 
 protected:
   void execTrajectory(const moveit_msgs::msg::RobotTrajectory& t) override;
 
 private:
-  ros::WallRate rate_;
+  rclcpp::WallRate rate_;
 };
 }
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controllers.h:129]:
```suggestion
```

To fix clang_format

**Comment by ddengster** [comment]
Alright, i've rebased and fixed the changes from the moveit1 sync. It's ready.

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp [hunk 16]
```diff
@@ -35,152 +35,156 @@
...
+    robot_model_loader::RobotModelLoader robot_model_loader(node_, ROBOT_DESCRIPTION);
+    const robot_model::RobotModelPtr& robot_model = robot_model_loader.getModel();
     moveit::core::RobotState robot_state(robot_model);
     typedef std::map<std::string, double> JointPoseMap;
     JointPoseMap joints;
 
     robot_state.setToDefaultValues();  // initialize all joint values (just in case...)
-    for (int i = 0, end = param.size(); i != end; ++i)
+
+    rcl_interfaces::msg::ListParametersResult params_result = node_->list_parameters({ param_base_name }, 2);
```
Comment (henningkayser) [inline moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp:152]: I actually thought that this should go the other way around, meaning we should declare the parameters for each group and then check if they have been specified. Eventually, we want to only support declared parameters if possible. Since we don't really have a best-practice for this yet, could you add a TODO to fix this in a follow-up?

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp [hunk 17]
```diff
@@ -35,152 +35,156 @@
...
         if (type == "last point")
-          controllers_[name].reset(new LastPointController(name, joints, pub_));
+          controllers_[controller_name].reset(new LastPointController(controller_name, controller_joints, pub_));
         else if (type == "via points")
-          controllers_[name].reset(new ViaPointController(name, joints, pub_));
+          controllers_[controller_name].reset(new ViaPointController(controller_name, controller_joints, pub_));
         else if (type == "interpolate")
-          controllers_[name].reset(new InterpolatingController(name, joints, pub_));
+        {
+          // @note: would put this in InterpolatingController's constructor
```
Comment (henningkayser) [inline moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp:111]: Why is this required now? If this is really only used in the `InterpolatingController` I would agree that we should move it there.

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp (hunk 4 above)
Comment (henningkayser) [reply moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp:60]: Yeah, I think that was related to adding support for configs outside of the private namespace. Since this is not an issue right now I would just remove this TODO. however, I would make the PARAM_BASE_NAME a `constexpr char PARAM_BASE_NAME[]` and make it global in this namespace.

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp (hunk 16 above)
Comment (ddengster) [reply moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp:153]: add todo 8af3088

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp (hunk 17 above)
Comment (ddengster) [reply moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp]: remove comment 0f02f39

File: moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp (hunk 4 above)
Comment (ddengster) [reply moveit_plugins/moveit_fake_controller_manager/src/moveit_fake_controller_manager.cpp]: remove todo 7a6b31d

**Comment by ddengster** [comment]
>  Any way we can test this?
There's no standalone test for it as far as I know. The only respite is that some of it was used in conjunction with PR #195. We'll have to wait for ports of the other modules to test it.
